src/librustc_trans/trans/adt.rs

```diff
@@ -466,23 +469,34 @@ struct Case<'tcx> {
 /// This represents the (GEP) indices to follow to get to the discriminant field
 pub type DiscrField = Vec<usize>;
 
-fn find_discr_field_candidate<'tcx>(tcx: &ty::ctxt<'tcx>,
+fn find_discr_field_candidate<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
                                     ty: Ty<'tcx>,
-                                    mut path: DiscrField) -> Option<DiscrField> {
+                                    mut path: DiscrField) -> Option<(DiscrField, ValueRef)> {
+    let ref tcx = cx.tcx();
     match ty.sty {
-        // Fat &T/&mut T/Box<T> i.e. T is [T], str, or Trait
-        ty::TyRef(_, ty::TypeAndMut { ty, .. }) | ty::TyBox(ty) if !type_is_sized(tcx, ty) => {
-            path.push(FAT_PTR_ADDR);
-            Some(path)
-        },
-
-        // Regular thin pointer: &T/&mut T/Box<T>
-        ty::TyRef(..) | ty::TyBox(..) => Some(path),
-
-        // Functions are just pointers
-        ty::TyBareFn(..) => Some(path),
+        // Fat &T/&mut T/Box<T> i.e. T is [T], str, or Trait: null is forbidden
+        ty::TyRef(_, ty::TypeAndMut { ty: ty2, .. }) | ty::TyBox(ty2)
+            if !type_is_sized(tcx, ty2) => {
+                path.push(FAT_PTR_ADDR);
+                Some((path, C_null(type_of::sizing_type_of(cx, ty))))
```

Comment:
I agree with the other comments provided regarding this line: as I understand it, you are trying to construct an instance of the forbidden value that will be used to tag whether we are in the unit-case or in the payload case.

For these fat `&Unsized` things, the code you have here will construct a null pointer of type `*struct { data: *i8, metadata: *i8 }`.  (This type description can be derived by inspecting the code for `fn sizing_type_of`, [here](https://github.com/rust-lang/rust/blob/master/src/librustc_trans/trans/type_of.rs#L189).)

There are other places in the code, namely `fn type_of` [here](https://github.com/rust-lang/rust/blob/master/src/librustc_trans/trans/type_of.rs#L277) where unsized pointers *do* have that doubly-indirect pointer shape (see comment above `fn type_of`). But in this case, if I understand correctly, the null in question that is being used to mark the unit case is just a simple `*i8` that will be stored in the first slot of that two word fat pointer.